### PR TITLE
Stop query execution if exception happened in PipelineExecutor itself.

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Stop query execution if exception happened in `PipelineExecutor` itself. This could prevent rare possible query hung. Continuation of #14334

### CODE REVIEW

File: src/Processors/Executors/PipelineExecutor.cpp [hunk 1]
```diff
@@ -432,6 +432,11 @@ bool PipelineExecutor::executeStep(std::atomic_bool * yield_flag)
         if (node->exception)
             std::rethrow_exception(node->exception);
 
+    /// Exception which happened in executing thread, but not at processor.
+    for (auto & executor_context : executor_contexts)
+        if (executor_context->exception)
+            std::rethrow_exception(executor_context->exception);
+
```
Comment (vitlibar) [inline src/Processors/Executors/PipelineExecutor.cpp:439]: It seems `executeStepImpl()` never set `executor_context->exception`, so this cycle doesn't look useful.

File: src/Processors/Executors/PipelineExecutor.cpp [hunk 2]
```diff
@@ -735,7 +731,16 @@ void PipelineExecutor::executeImpl(size_t num_threads)
                             CurrentThread::detachQueryIfNotDetached();
```
Comment (vitlibar) [inline src/Processors/Executors/PipelineExecutor.cpp:731]: It's not easy to understand the difference between these `execute*` functions.
Probably it's better to rename `executeImpl()` into something like `executeMultipleThreads()`?

File: src/Processors/Executors/PipelineExecutor.cpp (hunk 1 above)
Comment (KochetovNicolai) [reply src/Processors/Executors/PipelineExecutor.cpp]: Right. Wring place.

File: src/Processors/Executors/PipelineExecutor.cpp (hunk 2 above)
Comment (KochetovNicolai) [reply src/Processors/Executors/PipelineExecutor.cpp:731]: Agree.
Will fix it in other pr.

**Review by vitlibar — APPROVED** [review]

**Comment by KochetovNicolai** [comment]
`01412_cache_dictionary_race` is flappy.

**Pull request merged**